File: TarsEngine/Services/StyleIssueDetector.cs
using System.Text.RegularExpressions;
using Microsoft.Extensions.Logging;
using TarsEngine.Models;
using TarsEngine.Services.Interfaces;

namespace TarsEngine.Services;

/// <summary>
/// Detects style issues in code
/// </summary>
public class StyleIssueDetector(ILogger<StyleIssueDetector> logger) : IStyleIssueDetector
{
    private readonly ILogger<StyleIssueDetector> _logger = logger;

    /// <inheritdoc/>
    public string Language => "csharp";

    /// <inheritdoc/>
    public CodeIssueType IssueType => CodeIssueType.Style;

    /// <inheritdoc/>
    public List<CodeIssue> DetectIssues(string content, List<CodeStructure> structures)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            issues.AddRange(DetectInconsistentNaming(content));
            issues.AddRange(DetectInconsistentIndentation(content));
            issues.AddRange(DetectInconsistentBraceStyle(content));
            issues.AddRange(DetectMagicNumbers(content));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting style issues in code");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectInconsistentNaming(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Detect non-Pascal case class names
            var classRegex = new Regex(@"class\s+([a-zA-Z0-9_]+)", RegexOptions.Compiled);
            var classMatches = classRegex.Matches(content);
            foreach (Match match in classMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var className = match.Groups[1].Value;
                    if (!IsPascalCase(className))
                    {
                        var lineNumber = GetLineNumber(content, match.Index);
                        issues.Add(new CodeIssue
                        {
                            Type = CodeIssueType.Style,
                            Severity = TarsEngine.Models.IssueSeverity.Major,
                            Description = "Class name does not follow Pascal case convention",
                            Location = new CodeLocation
                            {
                                StartLine = lineNumber,
                                EndLine = lineNumber
                            },
                            CodeSnippet = className,
                            SuggestedFix = $"Rename to {ToPascalCase(className)}"
                        });
                    }
                }
            }

            // Detect non-Pascal case method names
            var methodRegex = new Regex(@"(public|private|protected|internal)?\s*(static|virtual|abstract|override|async)?\s*[a-zA-Z0-9_<>]+\s+([a-zA-Z0-9_]+)\s*\(", RegexOptions.Compiled);
            var methodMatches = methodRegex.Matches(content);
            foreach (Match match in methodMatches)
            {
                if (match.Groups.Count > 3)
                {
                    var methodName = match.Groups[3].Value;
                    if (!IsPascalCase(methodName))
                    {
                        var lineNumber = GetLineNumber(content, match.Index);
                        issues.Add(new CodeIssue
                        {
                            Type = CodeIssueType.Style,
                            Severity = TarsEngine.Models.IssueSeverity.Major,
                            Description = "Method name does not follow Pascal case convention",
                            Location = new CodeLocation
                            {
                                StartLine = lineNumber,
                                EndLine = lineNumber
                            },
                            CodeSnippet = methodName,
                            SuggestedFix = $"Rename to {ToPascalCase(methodName)}"
                        });
                    }
                }
            }

            // Detect non-camel case variable names
            var variableRegex = new Regex(@"(var|int|string|bool|double|float|decimal|char|byte|short|long|object|dynamic)\s+([a-zA-Z0-9_]+)\s*[=;]", RegexOptions.Compiled);
            var variableMatches = variableRegex.Matches(content);
            foreach (Match match in variableMatches)
            {
                if (match.Groups.Count > 2)
                {
                    var variableName = match.Groups[2].Value;
                    if (!IsCamelCase(variableName) && !variableName.StartsWith("_"))
                    {
                        var lineNumber = GetLineNumber(content, match.Index);
                        issues.Add(new CodeIssue
                        {
                            Type = CodeIssueType.Style,
                            Severity = TarsEngine.Models.IssueSeverity.Minor,
                            Description = "Variable name does not follow camel case convention",
                            Location = new CodeLocation
                            {
                                StartLine = lineNumber,
                                EndLine = lineNumber
                            },
                            CodeSnippet = variableName,
                            SuggestedFix = $"Rename to {ToCamelCase(variableName)}"
                        });
                    }
                }
            }

            // Detect non-underscore prefixed private fields
            var privateFieldRegex = new Regex(@"private\s+[a-zA-Z0-9_<>]+\s+([a-zA-Z0-9_]+)\s*[=;]", RegexOptions.Compiled);
            var privateFieldMatches = privateFieldRegex.Matches(content);
            foreach (Match match in privateFieldMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var fieldName = match.Groups[1].Value;
                    if (!fieldName.StartsWith("_"))
                    {
                        var lineNumber = GetLineNumber(content, match.Index);
                        issues.Add(new CodeIssue
                        {
                            Type = CodeIssueType.Style,
                            Severity = TarsEngine.Models.IssueSeverity.Minor,
                            Description = "Private field does not follow underscore prefix convention",
                            Location = new CodeLocation
                            {
                                StartLine = lineNumber,
                                EndLine = lineNumber
                            },
                            CodeSnippet = fieldName,
                            SuggestedFix = $"Rename to _{fieldName}"
                        });
                    }
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting inconsistent naming");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectInconsistentIndentation(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Split content into lines
            var lines = content.Split('\n');

            // Check for inconsistent indentation
            int expectedIndent = 0;
            int lineNumber = 1;

            foreach (var line in lines)
            {
                var trimmedLine = line.TrimStart();

                // Skip empty lines and comments
                if (string.IsNullOrWhiteSpace(trimmedLine) || trimmedLine.StartsWith("//") || trimmedLine.StartsWith("/*"))
                {
                    lineNumber++;
                    continue;
                }

                // Check if line starts with closing brace
                if (trimmedLine.StartsWith("}"))
                {
                    expectedIndent = Math.Max(0, expectedIndent - 4);
                }

                // Check indentation
                int actualIndent = line.Length - trimmedLine.Length;
                if (actualIndent != expectedIndent && !trimmedLine.StartsWith("#"))
                {
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Style,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = $"Inconsistent indentation (expected {expectedIndent} spaces, got {actualIndent})",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = line,
                        SuggestedFix = $"Indent with {expectedIndent} spaces"
                    });
                }

                // Check if line ends with opening brace
                if (trimmedLine.EndsWith("{"))
                {
                    expectedIndent += 4;
                }

                lineNumber++;
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting inconsistent indentation");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectInconsistentBraceStyle(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Split content into lines
            var lines = content.Split('\n');

            // Check for inconsistent brace style
            for (int i = 0; i < lines.Length - 1; i++)
            {
                var currentLine = lines[i].TrimEnd();
                var nextLine = lines[i + 1].TrimStart();

                // Check for K&R style (brace on same line)
                if ((currentLine.EndsWith("if (") || currentLine.EndsWith("else if (") ||
                     currentLine.EndsWith("for (") || currentLine.EndsWith("foreach (") ||
                     currentLine.EndsWith("while (") || currentLine.EndsWith("do") ||
                     currentLine.EndsWith("switch (")) &&
                    !currentLine.EndsWith("{"))
                {
                    // Find the closing parenthesis
                    int j = i + 1;
                    while (j < lines.Length && !lines[j].Contains(")"))
                    {
                        j++;
                    }

                    if (j < lines.Length)
                    {
                        var lineWithClosingParen = lines[j].TrimEnd();

                        // Check if the next line has the opening brace
                        if (j + 1 < lines.Length && lines[j + 1].TrimStart().StartsWith("{"))
                        {
                            issues.Add(new CodeIssue
                            {
                                Type = CodeIssueType.Style,
                                Severity = TarsEngine.Models.IssueSeverity.Minor,
                                Description = "Inconsistent brace style (Allman style detected)",
                                Location = new CodeLocation
                                {
                                    StartLine = j + 2,
                                    EndLine = j + 2
                                },
                                CodeSnippet = lines[j + 1],
                                SuggestedFix = "Use K&R style (opening brace on same line as control statement)"
                            });
                        }
                    }
                }

                // Check for Allman style (brace on new line)
                if (currentLine.EndsWith(")") && nextLine.StartsWith("{"))
                {
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Style,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = "Inconsistent brace style (Allman style detected)",
                        Location = new CodeLocation
                        {
                            StartLine = i + 2,
                            EndLine = i + 2
                        },
                        CodeSnippet = nextLine,
                        SuggestedFix = "Use K&R style (opening brace on same line as control statement)"
                    });
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting inconsistent brace style");
        }

        return issues;
    }

    /// <inheritdoc/>
    public List<CodeIssue> DetectMagicNumbers(string content)
    {
        var issues = new List<CodeIssue>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return issues;
            }

            // Detect magic numbers
            var magicNumberRegex = new Regex(@"[^a-zA-Z0-9_](\d+)[^a-zA-Z0-9_]", RegexOptions.Compiled);
            var magicNumberMatches = magicNumberRegex.Matches(content);

            foreach (Match match in magicNumberMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var number = match.Groups[1].Value;

                    // Skip common numbers like 0, 1, 2, -1
                    if (number == "0" || number == "1" || number == "2" || number == "-1" ||
                        number == "100" || number == "10" || number == "24" || number == "60" ||
                        number == "365" || number == "12" || number == "7" || number == "30" ||
                        number == "31")
                    {
                        continue;
                    }

                    // Skip numbers in array initializers
                    var context = GetContext(content, match.Index, 20);
                    if (context.Contains("new") && context.Contains("[") && context.Contains("]"))
                    {
                        continue;
                    }

                    // Skip numbers in enum declarations
                    if (context.Contains("enum") && context.Contains("{"))
                    {
                        continue;
                    }

                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Style,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = "Magic number detected",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = number,
                        SuggestedFix = "Replace with a named constant"
                    });
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error detecting magic numbers");
        }

        return issues;
    }

    /// <inheritdoc/>
    public Dictionary<TarsEngine.Services.Interfaces.IssueSeverity, string> GetAvailableSeverities()
    {
        return new Dictionary<TarsEngine.Services.Interfaces.IssueSeverity, string>
        {
            { TarsEngine.Services.Interfaces.IssueSeverity.Critical, "Critical style issue that violates team standards" },
            { TarsEngine.Services.Interfaces.IssueSeverity.Major, "High-impact style issue that affects readability significantly" },
            { TarsEngine.Services.Interfaces.IssueSeverity.Minor, "Medium-impact style issue" },
            { TarsEngine.Services.Interfaces.IssueSeverity.Trivial, "Low-impact style issue" },
            { TarsEngine.Services.Interfaces.IssueSeverity.Warning, "Informational style suggestion" }
        };
    }

    /// <inheritdoc/>
    public int GetLineNumber(string content, int position)
    {
        if (string.IsNullOrEmpty(content) || position < 0 || position >= content.Length)
        {
            return 0;
        }

        // Count newlines before the position
        return content[..position].Count(c => c == '\n') + 1;
    }

    /// <summary>
    /// Checks if a string is in Pascal case
    /// </summary>
    /// <param name="s">The string to check</param>
    /// <returns>True if the string is in Pascal case, false otherwise</returns>
    private static bool IsPascalCase(string s)
    {
        if (string.IsNullOrEmpty(s))
        {
            return false;
        }

        // First character must be uppercase
        if (!char.IsUpper(s[0]))
        {
            return false;
        }

        // Must not contain underscores
        if (s.Contains('_'))
        {
            return false;
        }

        return true;
    }

    /// <summary>
    /// Checks if a string is in camel case
    /// </summary>
    /// <param name="s">The string to check</param>
    /// <returns>True if the string is in camel case, false otherwise</returns>
    private static bool IsCamelCase(string s)
    {
        if (string.IsNullOrEmpty(s))
        {
            return false;
        }

        // First character must be lowercase
        if (!char.IsLower(s[0]))
        {
            return false;
        }

        // Must not contain underscores
        if (s.Contains('_'))
        {
            return false;
        }

        return true;
    }

    /// <summary>
    /// Converts a string to Pascal case
    /// </summary>
    /// <param name="s">The string to convert</param>
    /// <returns>The string in Pascal case</returns>
    private static string ToPascalCase(string s)
    {
        if (string.IsNullOrEmpty(s))
        {
            return s;
        }

        // Split by underscores and non-alphanumeric characters
        var parts = Regex.Split(s, @"[^a-zA-Z0-9]");

        // Convert each part to Pascal case
        for (int i = 0; i < parts.Length; i++)
        {
            if (!string.IsNullOrEmpty(parts[i]))
            {
                parts[i] = char.ToUpper(parts[i][0]) + (parts[i].Length > 1 ? parts[i][1..].ToLower() : "");
            }
        }

        return string.Join("", parts);
    }

    /// <summary>
    /// Converts a string to camel case
    /// </summary>
    /// <param name="s">The string to convert</param>
    /// <returns>The string in camel case</returns>
    private static string ToCamelCase(string s)
    {
        if (string.IsNullOrEmpty(s))
        {
            return s;
        }

        // Remove leading underscores
        s = s.TrimStart('_');

        // Convert to Pascal case first
        s = ToPascalCase(s);

        // Convert first character to lowercase
        return char.ToLower(s[0]) + s[1..];
    }

    /// <summary>
    /// Gets the context around a position in the content
    /// </summary>
    /// <param name="content">The content</param>
    /// <param name="position">The position</param>
    /// <param name="length">The length of context to get</param>
    /// <returns>The context</returns>
    private static string GetContext(string content, int position, int length)
    {
        if (string.IsNullOrEmpty(content) || position < 0 || position >= content.Length)
        {
            return string.Empty;
        }

        int start = Math.Max(0, position - length);
        int end = Math.Min(content.Length, position + length);

        return content.Substring(start, end - start);
    }
}

File: TarsEngine.FSharp.Core/Analysis/StyleIssueDetector.fs
namespace TarsEngine.FSharp.Core.Analysis

open System
open System.Text.RegularExpressions
open Microsoft.Extensions.Logging

/// <summary>
/// Implementation of IStyleIssueDetector for C# and F# languages.
/// </summary>
type StyleIssueDetector(logger: ILogger<StyleIssueDetector>) =
    
    /// <summary>
    /// Gets the language supported by this detector.
    /// </summary>
    member _.Language = "csharp"
    
    /// <summary>
    /// Detects issues in the provided content.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectIssues(content: string) =
        try
            logger.LogInformation("Detecting style issues")
            
            // Detect naming convention violations
            let namingConventionIssues = this.DetectNamingConventionViolations(content)
            
            // Detect formatting issues
            let formattingIssues = this.DetectFormattingIssues(content)
            
            // Detect code style inconsistencies
            let codeStyleIssues = this.DetectCodeStyleInconsistencies(content)
            
            // Detect comment style issues
            let commentStyleIssues = this.DetectCommentStyleIssues(content)
            
            // Combine all issues
            List.concat [
                namingConventionIssues
                formattingIssues
                codeStyleIssues
                commentStyleIssues
            ]
        with
        | ex ->
            logger.LogError(ex, "Error detecting style issues")
            []
    
    /// <summary>
    /// Detects issues in a file.
    /// </summary>
    /// <param name="filePath">The path to the file.</param>
    /// <returns>A list of detected issues.</returns>
    member this.DetectIssuesInFile(filePath: string) =
        try
            logger.LogInformation("Detecting style issues in file: {FilePath}", filePath)
            
            // Read the file content
            let content = System.IO.File.ReadAllText(filePath)
            
            // Detect issues in the content
            let issues = this.DetectIssues(content)
            
            // Update issues with file path
            issues |> List.map (fun issue -> { issue with FilePath = Some filePath })
        with
        | ex ->
            logger.LogError(ex, "Error detecting style issues in file: {FilePath}", filePath)
            []
    
    /// <summary>
    /// Gets the supported issue types.
    /// </summary>
    /// <returns>A list of supported issue types.</returns>
    member _.GetSupportedIssueTypes() =
        [
            CodeIssueType.Style
        ]
    
    /// <summary>
    /// Detects naming convention violations.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectNamingConventionViolations(content: string) =
        try
            // Define patterns for naming conventions
            let patterns = [
                // C# patterns
                @"class\s+([a-z][a-zA-Z0-9_]*)", "Class name should start with an uppercase letter"
                @"interface\s+([^I][a-zA-Z0-9_]*)", "Interface name should start with 'I'"
                @"enum\s+([a-z][a-zA-Z0-9_]*)", "Enum name should start with an uppercase letter"
                @"(public|private|protected|internal)\s+[a-zA-Z0-9_<>]+\s+([A-Z][a-zA-Z0-9_]*)\s*\(", "Method name should start with a lowercase letter"
                @"(public|private|protected|internal)\s+[a-zA-Z0-9_<>]+\s+([A-Z][a-zA-Z0-9_]*)\s*{", "Property name should start with a lowercase letter"
                @"(public|private|protected|internal)\s+[a-zA-Z0-9_<>]+\s+([A-Z][a-zA-Z0-9_]*)\s*;", "Field name should start with a lowercase letter"
                @"(public|private|protected|internal)\s+const\s+[a-zA-Z0-9_<>]+\s+([a-z][a-zA-Z0-9_]*)", "Constant name should be all uppercase"
                
                // F# patterns
                @"type\s+([a-z][a-zA-Z0-9_]*)", "Type name should start with an uppercase letter"
                @"type\s+([^I][a-zA-Z0-9_]*)\s*=\s*interface", "Interface name should start with 'I'"
                @"let\s+([A-Z][a-zA-Z0-9_]*)\s*=", "Value name should start with a lowercase letter"
                @"let\s+([A-Z][a-zA-Z0-9_]*)\s+[a-zA-Z0-9_]+", "Function name should start with a lowercase letter"
                @"member\s+(?:this|self|_)\.([A-Z][a-zA-Z0-9_]*)", "Member name should start with a lowercase letter"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    let name = m.Groups.[m.Groups.Count - 1].Value
                    
                    {
                        IssueType = CodeIssueType.Style
                        Severity = IssueSeverity.Info
                        Message = $"Naming convention violation: {message} ('{name}')"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Rename according to the naming convention"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting naming convention violations")
            []
    
    /// <summary>
    /// Detects formatting issues.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectFormattingIssues(content: string) =
        try
            // Define patterns for formatting issues
            let patterns = [
                // C# patterns
                @"if\s*\([^)]+\)[a-zA-Z0-9_]+", "Missing space after if statement"
                @"for\s*\([^)]+\)[a-zA-Z0-9_]+", "Missing space after for statement"
                @"while\s*\([^)]+\)[a-zA-Z0-9_]+", "Missing space after while statement"
                @"foreach\s*\([^)]+\)[a-zA-Z0-9_]+", "Missing space after foreach statement"
                @"if\s*\([^)]+\)\s*\{\s*\}", "Empty if block"
                @"for\s*\([^)]+\)\s*\{\s*\}", "Empty for block"
                @"while\s*\([^)]+\)\s*\{\s*\}", "Empty while block"
                @"foreach\s*\([^)]+\)\s*\{\s*\}", "Empty foreach block"
                @"catch\s*\([^)]+\)\s*\{\s*\}", "Empty catch block"
                @"finally\s*\{\s*\}", "Empty finally block"
                @"[ \t]+$", "Trailing whitespace"
                
                // F# patterns
                @"if\s+[^t]+then[a-zA-Z0-9_]+", "Missing space after then keyword"
                @"else[a-zA-Z0-9_]+", "Missing space after else keyword"
                @"for\s+[^i]+in[a-zA-Z0-9_]+", "Missing space after in keyword"
                @"while\s+[^d]+do[a-zA-Z0-9_]+", "Missing space after do keyword"
                @"match\s+[^w]+with[a-zA-Z0-9_]+", "Missing space after with keyword"
                @"[ \t]+$", "Trailing whitespace"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    {
                        IssueType = CodeIssueType.Style
                        Severity = IssueSeverity.Info
                        Message = $"Formatting issue: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Fix the formatting according to the style guide"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting formatting issues")
            []
    
    /// <summary>
    /// Detects code style inconsistencies.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectCodeStyleInconsistencies(content: string) =
        try
            // Define patterns for code style inconsistencies
            let patterns = [
                // C# patterns
                @"var\s+[a-zA-Z0-9_]+\s*=\s*new\s+([a-zA-Z0-9_<>]+)", "Explicit type can be inferred"
                @"([a-zA-Z0-9_<>]+)\s+[a-zA-Z0-9_]+\s*=\s*new\s+\1", "Type can be inferred using var"
                @"if\s*\([^)]+\)\s*return\s+[^;]+;\s*else\s*return\s+[^;]+;", "Can be simplified using conditional operator"
                @"if\s*\([^)]+\)\s*{\s*return\s+[^;]+;\s*}\s*else\s*{\s*return\s+[^;]+;\s*}", "Can be simplified using conditional operator"
                @"if\s*\([^)]+\)\s*{\s*[^}]+\s*}\s*else\s*{\s*[^}]+\s*}", "Consider using guard clauses"
                
                // F# patterns
                @"if\s+[^t]+then\s+[^e]+else\s+[^i]+", "Consider using pattern matching instead of if-then-else"
                @"match\s+[^w]+with\s*\|\s*[^-]+->\s*[^|]+\|\s*_\s*->\s*[^e]+", "Consider handling all cases explicitly instead of using wildcard pattern"
                @"let\s+[a-zA-Z0-9_]+\s*=\s*[^i]+\s+in\s+[^l]+", "Consider using let binding without in keyword"
                @"let\s+mutable\s+[a-zA-Z0-9_]+", "Consider using immutable values instead of mutable ones"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.Singleline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    {
                        IssueType = CodeIssueType.Style
                        Severity = IssueSeverity.Info
                        Message = $"Code style inconsistency: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Refactor the code according to the style guide"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting code style inconsistencies")
            []
    
    /// <summary>
    /// Detects comment style issues.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectCommentStyleIssues(content: string) =
        try
            // Define patterns for comment style issues
            let patterns = [
                // C# patterns
                @"//[a-zA-Z0-9]", "Missing space after comment delimiter"
                @"//\s*[a-z]", "Comment should start with an uppercase letter"
                @"//\s*[A-Z][^.!?]*[^.!?]$", "Comment should end with a period"
                @"/\*[a-zA-Z0-9]", "Missing space after comment delimiter"
                @"/\*\s*[a-z]", "Comment should start with an uppercase letter"
                @"[a-zA-Z0-9]\*/", "Missing space before comment delimiter"
                
                // F# patterns
                @"//[a-zA-Z0-9]", "Missing space after comment delimiter"
                @"//\s*[a-z]", "Comment should start with an uppercase letter"
                @"//\s*[A-Z][^.!?]*[^.!?]$", "Comment should end with a period"
                @"\(\*[a-zA-Z0-9]", "Missing space after comment delimiter"
                @"\(\*\s*[a-z]", "Comment should start with an uppercase letter"
                @"[a-zA-Z0-9]\*\)", "Missing space before comment delimiter"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    {
                        IssueType = CodeIssueType.Style
                        Severity = IssueSeverity.Info
                        Message = $"Comment style issue: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some "Fix the comment according to the style guide"
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting comment style issues")
            []
    
    interface IIssueDetector with
        member this.Language = this.Language
        member this.DetectIssues(content) = this.DetectIssues(content)
        member this.DetectIssuesInFile(filePath) = this.DetectIssuesInFile(filePath)
        member this.GetSupportedIssueTypes() = this.GetSupportedIssueTypes()
    
    interface IStyleIssueDetector with
        member this.DetectNamingConventionViolations(content) = this.DetectNamingConventionViolations(content)
        member this.DetectFormattingIssues(content) = this.DetectFormattingIssues(content)
        member this.DetectCodeStyleInconsistencies(content) = this.DetectCodeStyleInconsistencies(content)
        member this.DetectCommentStyleIssues(content) = this.DetectCommentStyleIssues(content)
